pkg/ebpf/events_derived.go

```diff
@@ -6,87 +6,102 @@ import (
+type deriveFn func(trace.Event) (trace.Event, bool, error)
 
-func (t *Tracee) deriveEvent(event trace.Event) []trace.Event {
-	var deriveFns map[int32]deriveFn
-	var derivatives []trace.Event
-
-	eventID := int32(event.EventID)
-	switch eventID {
-	case CgroupMkdirEventID:
-		containerCreateFn := func(def EventDefinition) error {
-			cgroupId, err := getEventArgUint64Val(&event, "cgroup_id")
-			if err != nil {
-				return err
-			}
+func (t *Tracee) initEventDerivationMap() error {
+	t.eventDerivations = map[int32]map[int32]deriveFn{
+		CgroupMkdirEventID: {
+			ContainerCreateEventID: deriveContainerCreate(t),
+		},
+		CgroupRmdirEventID: {
+			ContainerRemoveEventID: deriveContainerRemoved(t),
+		},
+	}
 
-			if info := t.containers.GetCgroupInfo(cgroupId); info.ContainerId != "" {
-				de := event
-				de.EventID = int(ContainerCreateEventID)
-				de.EventName = "container_create"
-				de.ReturnValue = 0
-				de.StackAddresses = make([]uint64, 1)
-				de.Args = []trace.Argument{
-					{ArgMeta: def.Params[0], Value: info.Runtime},
-					{ArgMeta: def.Params[1], Value: info.ContainerId},
-					{ArgMeta: def.Params[2], Value: info.Ctime.UnixNano()},
-				}
-				de.ArgsNum = 3
-
-				derivatives = append(derivatives, de)
-			}
+	return nil
+}
 
-			return nil
+func (t *Tracee) deriveEvent(event trace.Event) []trace.Event {
+	derivatives := []trace.Event{}
+	deriveFns := t.eventDerivations[int32(event.EventID)]
+	for id, deriveFn := range deriveFns {
+		// Don't derive events which were not requested by the user
+		if !t.eventsToTrace[id] {
+			continue
 		}
-		deriveFns = map[int32]deriveFn{
-			ContainerCreateEventID: containerCreateFn,
+
+		derivative, derived, err := deriveFn(event)
+		if err != nil {
+			t.handleError(fmt.Errorf("failed to derive event %d: %v", id, err))
+		} else if derived {
+			derivatives = append(derivatives, derivative)
 		}
-	case CgroupRmdirEventID:
-		containerRemoveFn := func(def EventDefinition) error {
-			cgroupId, err := getEventArgUint64Val(&event, "cgroup_id")
-			if err != nil {
-				return err
-			}
+	}
 
-			if info := t.containers.GetCgroupInfo(cgroupId); info.ContainerId != "" {
-				de := event
-				de.EventID = int(ContainerRemoveEventID)
-				de.EventName = "container_remove"
-				de.ReturnValue = 0
-				de.StackAddresses = make([]uint64, 1)
-				de.Args = []trace.Argument{
-					{ArgMeta: def.Params[0], Value: info.Runtime},
-					{ArgMeta: def.Params[1], Value: info.ContainerId},
-				}
-				de.ArgsNum = 2
-
-				derivatives = append(derivatives, de)
-			}
+	return derivatives
+}
 
-			return nil
+/*
+* Derivation functions:
+* Each derivation function initially takes it's runtime dependencies (usually tracee to track its state)
+* It then returns the actual deriveFn which is stored in tracee's derivation map (where it initializes the closure with its own state)
+* From the derivation map we call the relevant scoped deriveFn
```

Comment:
And we're limited to a single derived function per event type (which is okay as we're currently using one). I wonder if this comment is needed or it would be better to add description to each derive function AND `deriveEvent()` one, with a generic explanation.